src/v/cloud_storage/topic_manifest_downloader.cc

```diff
@@ -109,4 +112,106 @@ topic_manifest_downloader::download_manifest(
     co_return find_topic_manifest_outcome::no_matching_manifest;
 }
 
+namespace {
+using list_outcome_t
+  = std::vector<cloud_storage_clients::client::list_bucket_item>;
+ss::future<result<list_outcome_t, error_outcome>> find_prefixed_manifest_paths(
+  remote& remote,
+  cloud_storage_clients::bucket_name bucket,
+  retry_chain_node& parent_retry,
+  ss::lowres_clock::time_point deadline,
+  model::timestamp_clock::duration backoff,
+  const ss::sstring& hash_prefix) {
+    vassert(
+      hash_prefix.size() == 8,
+      "Expected prefix [0-9a-f]0000000, got {}",
+      hash_prefix);
+    retry_chain_node retry(deadline, backoff, &parent_retry);
+    auto prefixed_list_res = co_await remote.list_objects(
+      bucket,
+      retry,
+      cloud_storage_clients::object_key{fmt::format("{}/", hash_prefix)});
+    if (prefixed_list_res.has_error()) {
+        co_return error_outcome::manifest_download_error;
+    }
+    co_return prefixed_list_res.value().contents;
+}
+} // namespace
+
+ss::future<result<find_topic_manifest_outcome, error_outcome>>
+topic_manifest_downloader::find_manifests(
+  remote& remote,
+  cloud_storage_clients::bucket_name bucket,
+  retry_chain_node& parent_retry,
+  ss::lowres_clock::time_point deadline,
+  model::timestamp_clock::duration backoff,
+  std::optional<tp_ns_filter_t> tp_filter,
+  chunked_vector<topic_manifest>* manifests) {
+    retry_chain_node retry(deadline, backoff, &parent_retry);
+    absl::btree_set<model::topic_namespace> topics;
+    const auto maybe_add_topic =
+      [&topics,
+       &tp_filter](const std::optional<model::topic_namespace>& topic) {
+          if (!topic.has_value()) {
+              return;
+          }
+          if (!tp_filter.has_value() || tp_filter.value()(*topic)) {
+              topics.emplace(*topic);
+          }
+      };
+    {
+        // First, collect labeled topic manifests that match out filter.
+        const auto labeled_list_res = co_await remote.list_objects(
+          bucket,
+          retry,
+          cloud_storage_clients::object_key{
+            fmt::format("{}/", labeled_topic_manifests_root())});
+        if (labeled_list_res.has_error()) {
+            co_return error_outcome::manifest_download_error;
+        }
+        for (const auto& item : labeled_list_res.value().contents) {
+            auto tp_ns = from_labeled_topic_manifest_path(item.key);
+            maybe_add_topic(tp_ns);
+        }
+        // End scope to free memory.
+    }
+    // Next, collect prefixed topic manifests that match out filter.
+    const auto prefixed_roots = prefixed_topic_manifests_roots();
```

Comment:
+1 to Noah's comment about this function, but at the very least `static` here may help a little bit.

Reply:
Done